### PR TITLE
Display the message for last requests with foreign currency that are created in offline mode

### PR DESCRIPTION
**Pull request by dukenv0307**
### Fixed Issues

$ https://github.com/Expensify/App/issues/33073
PROPOSAL: https://github.com/Expensify/App/issues/33073#issuecomment-1857268114

### Tests

1. Go to staging.new.expensify.com.
2. Go offline
3.Click FAB > Request money > Manual.
4. Request money from a user that has no existing chat.
5. In 1:1 DM, click + > Request money.
6. Create another manual request in a different currency.
7. Click on the IOU preview.
8. Verify that: "Total will update when you're back online" will appear on the last IOU request with foreign currency.

- [x] Verify that no errors appear in the JS console

### QA Steps

1. Go to staging.new.expensify.com.
2. Go offline
3.Click FAB > Request money > Manual.
4. Request money from a user that has no existing chat.
5. In 1:1 DM, click + > Request money.
6. Create another manual request in a different currency.
7. Click on the IOU preview.
8. Verify that: "Total will update when you're back online" will appear on the last IOU request with foreign currency.

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/129500732/3f93727d-d1d6-4a2c-bfce-4ee7e0547933

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/129500732/132e55b5-b9e5-4a3d-b811-9da58a4095da

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/129500732/4edd0317-1aba-4301-a66b-3ccfe8580b8a

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/129500732/f0db44ac-a8f7-4cd9-a361-4e988e85fb3f

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/129500732/63201b18-214a-42cf-82e6-fcac89e6f8e3

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/129500732/562c9dd2-5af3-487a-9562-c59077943def

</details>

### CODE REVIEW

**Comment by cubuspl42** [comment]
I started [a discussion](https://expensify.slack.com/archives/C01GTK53T8Q/p1704456536304589) on Slack, as this is a slight change of a user-observable behavior. I think it's a minor one, so we should be good, but it's better to confirm it.

**Comment by cubuspl42** [comment]
No replies; let's just move forward. This is an edge case.

**Comment by cubuspl42** [comment]
Please merge `main`, while I'll proceed with the testing

**Comment by cubuspl42** [comment]
Hey, I'm sorry, but I think we'll have to pivot. The solution works different form what I understood earlier.

I thought that we'll display the "Total will update..." under each entry which _isn't included in the total_. That made sense to me.

In practice, we also display the mentioned message under entires which _are_ included in the total.

https://github.com/Expensify/App/assets/2590174/33f71622-41cb-484f-bfbc-202fdea17f95

I might've misunderstood "to display the message for all requests with foreign currency that are created in offline mode". I thought that it means all non-PLN requests (in my example).

**Comment by cubuspl42** [comment]
> @cubuspl42 Updated.

Thank you for your update. Please note that very short responses make me unsure whether we're on the same page. The vast majority of the questions I ask on PRs are not rhetorical, i.e., I actually expect an answer. I'd prefer to know whether I misunderstood a detail in the proposal, or if I got it right, but the implementation had a bug that made the code not equivalent to the original idea.

File: src/components/ReportActionItem/MoneyRequestAction.js [hunk 1]
```diff
@@ -119,15 +115,8 @@ function MoneyRequestAction({
...
-        !_.isEmpty(iouReport) &&
-        !_.isEmpty(reportActions) &&
-        chatReport.iouReportID &&
-        isMostRecentIOUReportAction &&
-        action.pendingAction === CONST.RED_BRICK_ROAD_PENDING_ACTION.ADD &&
-        network.isOffline
-    ) {
-        shouldShowPendingConversionMessage = IOUUtils.isIOUReportPendingCurrencyConversion(iouReport);
+    if (!_.isEmpty(iouReport) && !_.isEmpty(reportActions) && chatReport.iouReportID && action.pendingAction === CONST.RED_BRICK_ROAD_PENDING_ACTION.ADD && network.isOffline) {
+        shouldShowPendingConversionMessage = IOUUtils.isTransactionPendingCurrencyConversion(iouReport, (action && action.originalMessage && action.originalMessage.IOUTransactionID) || 0);
```
Comment (cubuspl42) [inline src/components/ReportActionItem/MoneyRequestAction.js:119]: Commenting on `|| 0`...

The relevant function argument is typed: `transactionID: string`

So I think that either the argument type or the passed value is not correct

**Comment by cubuspl42** [comment]
I'm sorry about the back-and-forth, but I'd like to confirm one thing.

I never had strong opinions about this issue. I just want a simple fix with reasonable behavior.

In your proposal, you sketched two options:

1.
  - > Pass the correct chatReportID here [...]
  - > Use useMemo here [...]
  
2. ...or...
  - > we can remove this check here to display the message for all requests with **foreign** currency that are created in offline mode
  
Option 2 sounded like a cool idea that is _very_ easy to implement and has resonable behavior.

I suggested "cool, let's go with option 2 then", as I thought that we're gonna remove one check and fix the issue this way.

At this point, it doesn't seem like it was removing one expression.

What is your perspective on pivoting back to Option 1? Do I understand correctly, that it would be less invasive, both in the user-perceived behavior and the amount of modified code?

**Comment by cubuspl42** [comment]
> What is your perspective on pivoting back to Option 1? Do I understand correctly, that it would be less invasive, both in the user-perceived behavior and the amount of modified code?

Again, this wasn't a rhetorical question! 🙂 I always want to know the Contributor's perspective on my suggestions. If they disagree, it can go both ways: either I accept their perspective or insist if I believe that it's necessary.

**Comment by cubuspl42** [comment]
@dukenv0307 

About the commit message: I'm 90% sure that this one won't do any harm (as it's missing the `#`), but I would just recommend against any PR/commit descriptions involving "Fix {issue_id}". In this case, it could be whatever else, even "Fix issue with foreign currencies".

For now, you can leave it as is, but let's have this in mind.

**Comment by cubuspl42** [comment]
Please change the PR title

**Comment by cubuspl42** [comment]
In the Tests let's specify that the message should appear on the last entry. We pivoted our approach, it's not good that the tests still "pass".

**Comment by cubuspl42** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos

<details>
<summary>Android: Native</summary>

image

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/2590174/079efd5e-1e6f-499d-b121-d906e391b4a9

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/2590174/9eccafc6-1aaa-469f-bb2d-a44a0a88ff06

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/2590174/f03fb0d2-64a7-4527-99a5-bc81148bd34e

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/2590174/086213e8-a2a6-4ca1-9bcf-c4c39fde89bf

</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Review by jakub-trzebiatowski — APPROVED** [review]

File: src/pages/home/report/ReportActionItem.js [hunk 2]
```diff
@@ -346,7 +346,7 @@ function ReportActionItem(props) {
             const iouReportID = originalMessage.IOUReportID ? originalMessage.IOUReportID.toString() : '0';
             children = (
                 <MoneyRequestAction
-                    chatReportID={props.report.reportID}
+                    chatReportID={originalMessage.IOUReportID ? props.report.chatReportID : props.report.reportID}
```
Comment (arosiclair) [inline src/pages/home/report/ReportActionItem.js:349]: Can you just add a comment explaining this check so the next person knows why we needed to do this?

**Review by arosiclair — CHANGES_REQUESTED** [review]

File: src/pages/home/report/ReportActionItem.js (hunk 2 above)
Comment (cubuspl42) [reply src/pages/home/report/ReportActionItem.js:351]: Please ensure that this is in proper English. You can use freely available tools like ChatGPT 3.

Comment (arosiclair) [reply src/pages/home/report/ReportActionItem.js:351]: Try this

> If originalMessage.iouReportID is set, this is a 1:1 money request in a DM chat whose reportID is props.report.chatReportID

**Review by arosiclair — APPROVED** [review]

**Comment by arosiclair** [comment]
Unrelated error in the performance test. I re-ran it hopefully it passes now.

**Comment by cubuspl42** [comment]
@arosiclair Checks passing 👍

**Pull request merged**